php/libraries/User.class.inc

```diff
@@ -411,89 +422,106 @@ class User extends UserPermissions
-        // (ie the second value/operator pair is always the UserID of
-        // the user whose password is being checked, the third is their
-        // email, and the first is the "Confirm" password which has
-        // nothing to do with whether the password is strong or not..)
-        $CharTypes = 0;
-        // less than 8 characters
-        if (strlen($password) < 8) {
-            return false;
-        }
-
-        // nothing but letters
-        if (!preg_match('/[^A-Za-z]/', $password)) {
-            return false;
-        }
-
-        // nothing but numbers
-        if (!preg_match('/[^0-9]/', $password)) {
-            return false;
-        }
+    function checkPasswordStrength(string $password): array
+    {
+        $user = \User::singleton();
+        $data = $user->getData();
+        /* This is an array of information related to the user. This is used by
+         * Zxcvbn to check if a user is doing something foolish like making
+         * their password 'johnnyloris'.
+         * $lorisKeywords is a list that reduce the overall score of a potential
+         * password. It comes from the file docs/LorisKeywords. Developers
+         * are encouraged to add relevant words to this file relating to their
+         * projects.
+         * A useful tool for this tasks can be found at the link below.
+         * https://tools.kali.org/password-attacks/cewl
+         */
+        $lorisKeywords = self::_getLorisKeywords();
+        $userInputs    = array_merge(
+            array(
+             $user->getUsername(),
+             $data['Email'],
+             $data['Real_name'],
+            ),
+            $lorisKeywords
+        );
 
-        // preg_match returns 1 on match, 0 on non-match
-        $CharTypes += preg_match('/[0-9]+/', $password);
-        $CharTypes += preg_match('/[A-Za-z]+/', $password);
-        $CharTypes += preg_match('/[!\\\$\^@#%&\*\(\)]+/', $password);
-        if ($CharTypes < 3) {
-            return false;
-        }
+        $zxcvbn = new Zxcvbn();
+        return $zxcvbn->passwordStrength($password, $userInputs);
+    }
 
-        // compare password to values
-        foreach ($values as $key => $value) {
-            switch ($operators[$key]) {
-            case '==':
-                // They must be equal, so the whole thing is false
-                // if they're not.
-                if ($password != $value) {
-                    return false;
-                }
-                break;
-            case '!=':
-                if ($password == $value) {
-                    return false;
+    /**
+     * Slurps contents of keyword file and returns a word list.
+     *
+     * @return array Keywords that shouldn't be used in passwords
+     */
+    private function _getLorisKeywords(): array
+    {
+        $keywords = array();
+        // The file is located two directories up from this file.
+        $keywordFile = __DIR__ . '/../../' . 'docs/LorisKeywords';
```

Comment:
I don't think this is a very robust way of loading the file or a good place to put a config file that users might customize.

Can we move the file to `$PROJECT_DIR/password_blacklist.txt` (or something like that?)

Reply:
(and put the default there on install)

Reply:
To be clear you mean something like "/var/www/loris/" and not the `project/` dir, right?

Reply:
no, I mean it should be in the `project` directory since the project is going to customize it.

Reply:
@ridz1208 I can reconfigure this PR to use a new table instead.

Reply:
@driusan Do you have a preference here? At this time the text file serves as both a password blacklist and as a list of plaintext words that should reduce a password's score (but not necessarily cause it to fail).

Reply:
A table is probably more efficient for large blacklists